packages/oid4vci-holder/src/agent/OID4VCIHolder.ts

```diff
@@ -0,0 +1,289 @@
+  private readonly onCredentialStored?: (args: OnCredentialStoredArgs) => Promise<void>
+  private readonly onGetCredentials: (args: OnGetCredentialsArgs) => Promise<Array<CredentialToAccept>>
+
+  constructor(options: OID4VCIHolderOptions) {
+    const { onContactIdentityCreated, onCredentialStored, onGetCredentials, vcFormatPreferences } = options
+
+    if (vcFormatPreferences !== undefined && vcFormatPreferences.length > 0) {
+      this.vcFormatPreferences = vcFormatPreferences
+    }
+    this.onGetCredentials = onGetCredentials
+    this.onContactIdentityCreated = onContactIdentityCreated
+    this.onCredentialStored = onCredentialStored
+  }
+
+  public async onEvent(event: any, context: RequiredContext): Promise<void> {
+    switch (event.type) {
+      case OID4VCIHolderEvent.CONTACT_IDENTITY_CREATED:
+        this.onContactIdentityCreated?.(event.data)
+        break
+      case OID4VCIHolderEvent.CREDENTIAL_STORED:
+        this.onCredentialStored?.(event.data)
+        break
+      default:
+        return Promise.reject(Error('Event type not supported'))
+    }
+  }
+
+  private async oid4vciHolderGetMachineInterpreter(args: OID4VCIMachineInstanceOpts, context: RequiredContext): Promise<OID4VCIMachineInterpreter> {
+    const services = {
+      initiateOID4VCI: (args: InitiateOID4VCIArgs) => this.oid4vciHolderGetInitiationData(args, context),
+      createCredentialSelection: (args: CreateCredentialSelectionArgs) => this.oid4vciHolderCreateCredentialSelection(args, context),
+      getContact: (args: GetContactArgs) => this.oid4vciHolderGetContact(args, context),
+      getCredentials: (args: GetCredentialsArgs) => this.oid4vciHolderGetCredentials(args, context),
+      addContactIdentity: (args: AddContactIdentityArgs) => this.oid4vciHolderAddContactIdentity(args, context),
+      assertValidCredentials: (args: AssertValidCredentialsArgs) => this.oid4vciHolderAssertValidCredentials(args, context),
+      storeCredentialBranding: (args: StoreCredentialBrandingArgs) => this.oid4vciHolderStoreCredentialBranding(args, context),
+      storeCredentials: (args: StoreCredentialsArgs) => this.oid4vciHolderStoreCredentials(args, context),
+    }
+
+    const oid4vciMachineInstanceArgs: OID4VCIMachineInstanceOpts = {
+      ...args,
+      services: {
+        ...services,
+        ...args.services,
+      },
+    }
+
+    return OID4VCIMachine.newInstance(oid4vciMachineInstanceArgs)
+  }
+
+  private async oid4vciHolderGetInitiationData(args: InitiateOID4VCIArgs, context: RequiredContext): Promise<InitiationData> {
+    const { requestData } = args
+
+    if (requestData?.uri === undefined) {
+      return Promise.reject(Error('Missing request uri in context'))
+    }
+
+    if (
+      !requestData?.uri ||
+      !(requestData?.uri.startsWith(RequestType.OPENID_INITIATE_ISSUANCE) || requestData?.uri.startsWith(RequestType.OPENID_CREDENTIAL_OFFER))
+    ) {
+      return Promise.reject(Error('Invalid Uri'))
```

Comment:
Please provide a bit more context, If you want to keep messages to a user clean we will need to device  an error probably that can have more context for logging and a simple message for users. Or we do the logging here, but that is a bit of an anti-pattern.

In this case it would be very helpful to have the actual URI in the logs in case it goes wrong

Reply:
added the uri